src/backend/distributed/worker/worker_data_fetch_protocol.c

```diff
@@ -662,17 +664,30 @@ worker_append_table_to_shard(PG_FUNCTION_ARGS)
 	/* make sure we are allowed to execute the COPY command */
 	CheckCopyPermissions(localCopyCommand);
 
-	/* need superuser to copy from files */
-	GetUserIdAndSecContext(&savedUserId, &savedSecurityContext);
-	SetUserIdAndSecContext(CitusExtensionOwner(), SECURITY_LOCAL_USERID_CHANGE);
+	Relation shardRelation = table_openrv(localCopyCommand->relation, RowExclusiveLock);
+
+	/* mimic check from copy.c */
+	if (XactReadOnly && !shardRelation->rd_islocaltemp)
```

Comment:
In general, what is the difference between using `DoCopy` vs `BeginCopy&CopyFrom&EndCopy`? 

Can we use `DoCopy` so that we also apply all the other checks?

Reply:
DoCopy also includes the superuser checks that we're trying to avoid.